python/ShapeTools.py

```diff
@@ -65,16 +69,63 @@ def doIndividualModels(self):
                 pdfs.add(pdf); coeffs.add(coeff)
                 if not self.DC.isSignal[p]:
                     bgpdfs.add(pdf); bgcoeffs.add(coeff)
+                else:
+                    sigcoeffs.append(coeff)
             if self.options.verbose > 1: print "Creating RooAddPdf %s with %s elements" % ("pdf_bin"+b, coeffs.getSize())
-            sum_s = ROOT.RooAddPdf("pdf_bin%s"       % b, "",   pdfs,   coeffs)
+            if self.options.newHist >= 1:
+                prop = ROOT.CMSHistErrorPropagator("prop_bin%s" % b, "", self.out.binVar, pdfs, coeffs)
+                if self.options.newHistBinPars >= 0.:
+                    bbb_args = prop.setupBinPars(self.options.newHistBinPars)
+                    # bbb_args.Print()
+                    for bidx in range(bbb_args.getSize()):
+                        arg = bbb_args.at(bidx)
+                        n = arg.GetName()
+                        self.out._import(arg)
+                        if arg.getAttribute("createGaussianConstraint"):
+                            self.doObj("%s_Pdf" % n, "SimpleGaussianConstraint", "%s, %s_In[0,%s], %s" % (n, n, '-7,7', '1.0'), True)
+                            self.out.var(n).setVal(0)
+                            self.out.var(n).setError(1)
+                            if self.options.optimizeBoundNuisances: self.out.var(n).setAttribute("optimizeBounds")
+                        elif arg.getAttribute("createPoissonConstraint"):
+                            nom = arg.getVal()
+                            pval = ROOT.Math.normal_cdf_c(7)
+                            minObs = nom
+                            while minObs > 0 and (ROOT.TMath.Poisson(minObs, nom + 1) > pval):
+                                minObs -= (sqrt(nom) if nom > 10 else 1)
+                            maxObs = nom + 2
+                            while (ROOT.TMath.Poisson(maxObs, nom + 1) > pval):
+                                #print "Poisson(maxObs = %d, %f) = %g > 1e-12" % (maxObs, args[0]+1, ROOT.TMath.Poisson(maxObs, args[0]+1))
+                                maxObs += (sqrt(nom) if nom > 10 else 2)
+                            self.doObj("%s_Pdf" % n, "Poisson", "%s_In[%d,%f,%f], %s, 1" % (n, nom, minObs, maxObs, n))
+                        binconstraints.add(self.out.pdf('%s_Pdf' % n))
+                        self.out.var("%s_In" % n).setConstant(True)
+                        self.extraNuisances.append(self.out.var("%s" % n))
+                        self.extraGlobalObservables.append(self.out.var("%s_In" % n))
+                if not self.out.var('ONE'):
+                    self.doVar('ONE[1.0]')
+                sum_s = ROOT.RooRealSumPdf("pdf_bin%s"       % b,  "", ROOT.RooArgList(prop),   ROOT.RooArgList(self.out.var('ONE')), True)
+                if not self.options.noBOnly:
+                    if not self.out.var('ZERO'):
+                        self.doVar('ZERO[0.0]')
+                    customizer = ROOT.RooCustomizer(prop, "")
```

Comment:
The definition of the b-only model is a bit different here. We could have created a separate HistErrorPropagator with only background processes, but then setupBinPars could produce a non-overlapping set of bin parameters compared to the s+b model. It's simpler to clone the s+b HistErrorPropagator and force all signal norm terms to zero instead.